_posts/2023-03-11-Adding-a-Timeout-feature-to-Rustgem.md

```diff
@@ -0,0 +1,288 @@
+callback when the fence is not signaled by the user for more than 10 seconds.
+Gladly, the Linux kernel provides us with a Timer that can be set with a
+callback and a timeout. But, to use it in the Rust code, we need to have a safe
+abstraction, that will ensure that the code is safe under some assumptions.
+
+# First Attempt: writing a Timer abstraction
+---
+
+So, initially, I was developing an abstraction on my own as I checked the RfL
+tree and there were no Timer abstractions available.
+
+The most important question here is "how can we guarantee access to other
+objects inside the callback?". The callback only has, as parameters, a pointer
+to the `struct timer_list`. Naturally, we can think about using a `container_of`
+macro. So, in order to make the compatibility layer between Rust and the C
+callback, I decided to store the object inside the Timer. Yep, I didn't like
+that a lot, but it was the solution I came in at the time. So, the struct looked
+something like this:
+
+```rust
+/// A driver-specific Timer Object
+//
+// # Invariants
+// timer is a valid pointer to a struct timer_list and we own a reference to it.
+[repr(C)]
+pub struct UniqueTimer<T: TimerOps<Inner = D>, D> {
+   timer: *bindings::timer_list,
+   inner: D,
+   _p: PhantomData<T>,
+}
+```
+
+Moreover, the second important question I had was "how can the user pass a
+callback function to the timer?". There were two possibilities: using a closure
+and using a Trait. I decided to go through the trait path. Things would be kind
+of similar if I decided to go into the closure path.
+
+```rust
+/// Trait which must be implemented by driver-specific timer objects.
+pub trait TimerOps: Sized {
+    /// Type of the Inner data inside the Timer
+    type Inner;
+
+    /// Timer callback
+    fn timer_callback(timer: &UniqueTimer<Self, Self::Inner>);
+}
+```
+
+With those two questions solved, it seems that we are all set and good to go.
+So, we can create methods to initialize the timer and modify the timer's
+timeout, implement the Drop trait, and use the following callback by default:
+
+```rust
+unsafe extern "C" fn timer_callback<T: TimerOps<Inner = D>, D: Sized>(
+    timer: *mut bindings::timer_list,
+) {
+    let timer = crate::container_of!(timer, UniqueTimer<T, D>, timer)
+			    as *mut UniqueTimer<T, D>;
+
+    // SAFETY: The caller is responsible for passing a valid timer_list subtype
+    T::timer_callback(unsafe { &mut *timer });
+}
+```
+
+All should work, right? Well... No, I didn't really mention how I was allocating
+memory. And let's say I was initially allocating it wrongly and therefore, the
+`container_of` macro was pointing to the wrong memory space.
+
+Initially, I was allocating only `timer` with the kernel memory allocator
+`krealloc` and allocating the rest of the struct with Rust's memory allocator.
+By making such a mess, `container_of` wasn't able to point to the right
+direction.
```

Comment:
I found this a bit confusing. "Direction" is the memory address? Or is the word used in a general meaning?